{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys\n",
    "from __future__ import print_function\n",
    "import cPickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with open(\"../reads/meta/s2.tsv\") as f:\n",
    "    data = pd.read_table(f).set_index('EMBL ID').drop([\"Strain/species details\"], axis=1).to_dict()[\"Phylum\"]\n",
    "id2phlm = {}\n",
    "for k,v in data.items():\n",
    "    nk = k.split(\".\")[0]\n",
    "    id2phlm[nk] = v\n",
    "    if (nk == \"CM000636\"):\n",
    "        id2phlm[\"CP006835\"] = v\n",
    "del data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Done reading 6 Million reads.."
     ]
    }
   ],
   "source": [
    "mil = 1000000\n",
    "hund = 100\n",
    "\n",
    "def get_ref_id(aln):\n",
    "    # get id for the mapped reference\n",
    "    try:\n",
    "        return aln.reference_name.split(\"|\")[1]\n",
    "    except:\n",
    "        return aln.reference_name.split(\".\")[0]\n",
    "    \n",
    "def print_details(qId, rId, aln):\n",
    "    print (\"ERROR\")\n",
    "    print(\"SAM Algns\")\n",
    "    print (aln.query_name, aln.reference_name)\n",
    "    print(\"SUBSETTING\")\n",
    "    print (qId, rId)\n",
    "    print (\"PHYLA\")\n",
    "    print (id2phlm[qId], id2phlm[rId])\n",
    "\n",
    "\n",
    "def get_stats(fname, rname, level=\"phyla\"):\n",
    "    totReads = 0\n",
    "    TN = 0\n",
    "    FN = 0\n",
    "    reads_list = []\n",
    "    roseCount = 0\n",
    "    euCount = 0\n",
    "    with open(rname) as f:        \n",
    "        for line in f:\n",
    "            # counting total reads\n",
    "            totReads += 1\n",
    "            # Progress Monitoring\n",
    "            if(totReads % mil == 0):\n",
    "                print (\"\\r Done reading {} Million reads.\".format(int(round(totReads)/1000000)), end=\"\")\n",
    "            \n",
    "            #extracting relevant part of read\n",
    "            read = line.strip().replace(\"/1\",\"\").replace(\"@\",\"\")\n",
    "            \n",
    "            if \"Random\" in read:\n",
    "                TN += 1\n",
    "            if \"Eukaryotes\" in read:\n",
    "                euCount += 1\n",
    "            if \"Rose\" in read:\n",
    "                roseCount += 1\n",
    "            \n",
    "            #making a list of read id\n",
    "            reads_list.append(read)\n",
    "            \n",
    "            # skip next 4 lines\n",
    "            for _ in range(3):\n",
    "                f.next()\n",
    "    \n",
    "    if len(reads_list) != len(set(reads_list)):\n",
    "        print (\"ERROR: Repeating reads found\")\n",
    "        return 0\n",
    "\n",
    "# ++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
    "# Repeating Block\n",
    "# ++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
    "#     print(\"\\nSaving Pickle\")\n",
    "#     with open(r\"reads_list.pickle\", \"wb\") as f:\n",
    "#         cPickle.dump(reads_list, f)\n",
    "#     print(\"Done Saving Pickle\")\n",
    "# ++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
    "\n",
    "# ++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
    "# Repeating Block\n",
    "# ++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
    "#     print(\"Reading Pickle\")\n",
    "#     with open(r\"reads_list.pickle\", \"rb\") as f:\n",
    "#         reads_list = cPickle.load(f)\n",
    "#     totReads = len(reads_list)\n",
    "# ++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
    "    \n",
    "    with pysam.AlignmentFile(fname) as f:\n",
    "        TP = 0\n",
    "        FP = 0\n",
    "        totCount = 0.0\n",
    "        singCount = 0\n",
    "        orphanCount = 0\n",
    "        skipCount = 0.0\n",
    "        for aln in f:\n",
    "            #get mate of the read\n",
    "            mate_aln = f.next()\n",
    "            \n",
    "            # count total Number of reads\n",
    "            totCount += 1\n",
    "            \n",
    "            # get number of alignments\n",
    "            n_alns = aln.get_tag('NH')\n",
    "            \n",
    "            #ignoring Rose Sequence\n",
    "            if \"Rose\" in aln.query_name or \"Eukaryotes\" in aln.query_name:\n",
    "                skipCount += 1.0/n_alns\n",
    "                continue\n",
    "            \n",
    "            # Ignoring Orphan alignments for now\n",
    "            if(aln.reference_name != mate_aln.reference_name):\n",
    "                orphanCount += 1\n",
    "                print (\"WARNING: ORPHANS Detected statistics Neess to be re-evaluated\")\n",
    "                continue\n",
    "            \n",
    "            # Progress Monitoring\n",
    "            if(round(totCount) % mil == 0):\n",
    "                print (\"\\r Done reading {} Million reads.\".format(int(round(totCount)/1000000)), end=\"\")\n",
    "\n",
    "            # get ground truth id\n",
    "            qId = aln.query_name.split('-')[0]\n",
    "            if \"|\" in qId:\n",
    "                qId = qId.split(\"|\")[1]\n",
    "            elif \"_\" in qId:\n",
    "                qId = qId.split(\"_\")[0]\n",
    "                \n",
    "            # for singly mapped reads only\n",
    "            if n_alns == 1:\n",
    "                # Increment the single count\n",
    "                singCount += 1\n",
    "\n",
    "                # get id for the mapped reference\n",
    "                rId = get_ref_id(aln)\n",
    "                \n",
    "                # compare the labels\n",
    "                try:\n",
    "                    if qId == rId or id2phlm[qId]==id2phlm[rId]:\n",
    "                        # get TP count \n",
    "                        TP += 1\n",
    "                    else:\n",
    "#                         print_details(qId, rId, aln)\n",
    "                        # get FP count \n",
    "                        FP += 1\n",
    "                except:\n",
    "                    print (\"In Single mapping\")\n",
    "                    print_details(qId, rId, aln)\n",
    "\n",
    "                    \n",
    "            #handling MultiMapped Reads\n",
    "            else:\n",
    "                # list of all alignments\n",
    "                algns = [get_ref_id(aln)]\n",
    "                \n",
    "                # iterate over all alignments\n",
    "                for _ in range(1, n_alns):\n",
    "                    aln = f.next()\n",
    "                    mate_aln = f.next()\n",
    "                    \n",
    "                    # Ignoring Orphan alignments for now\n",
    "                    if(aln.reference_name != mate_aln.reference_name):\n",
    "                        orphanCount += 1\n",
    "                    else:\n",
    "                        algns.append(get_ref_id(aln))\n",
    "                \n",
    "                # SIMULATED FALSE POSITIVES\n",
    "                if \"Random\" in aln.query_name:\n",
    "                    TN -= 1\n",
    "                    FP += 1\n",
    "                    continue\n",
    "                \n",
    "                flag = False\n",
    "                for rId in algns:\n",
    "                    try:\n",
    "                        if qId == rId or id2phlm[qId]==id2phlm[rId]:\n",
    "                            flag = True\n",
    "                            break\n",
    "                    except:\n",
    "                        print (\"In Multimapping\")\n",
    "                        print_details(qId, rId, aln)\n",
    "\n",
    "                if flag:\n",
    "                    TP += 1\n",
    "                else:\n",
    "#                     print_details(qId, rId, aln)\n",
    "                    FP += 1\n",
    "    \n",
    "    mmCount = round(totCount - singCount)\n",
    "    unmapCount = totReads - totCount\n",
    "    FN = totReads - TP - FP - TN - roseCount - euCount\n",
    "    print (\"\\n\")\n",
    "    print (\"====================================================================================\")\n",
    "    print (\"Total Number of reads: {0} ({1:.2f}M)\".format(totReads, totReads/mil))    \n",
    "    print (\"Number of Unmapped reads: {0} ({1:.2f}M, {2:.2f}%)\".format(unmapCount, unmapCount/mil, unmapCount*hund/totReads))    \n",
    "    print (\"Number of Mapped reads {0}({1:.2f}M, {2:.2f}%)\".format(totCount, totCount/mil, totCount*hund/totReads))\n",
    "    print (\"\\n\")\n",
    "    print (\"============================ OUT OF MAPPED READS ===================================\")\n",
    "    print (\"Number of Singly Mapped reads: {0} ({1:.2f}M, {2:.2f}%)\".format(singCount, singCount/mil, singCount*hund/totReads))\n",
    "    print (\"Number of Multimapped reads: {0} ({1:.2f}M, {2:.2f}%)\".format(mmCount, mmCount/mil, mmCount*hund/totReads))\n",
    "    print (\"Number of Mapped but Skipped reads: {0} ({1:.2f}M, {2:.2f}%)\".format(skipCount, skipCount/mil, skipCount*hund/totReads))\n",
    "    print (\"Number of Orphaned (Ignored)ALIGNMENTS (Should be significantly low): {}\".format(orphanCount))\n",
    "    print (\"====================================================================================\")\n",
    "    print (\"\\n ===================== \\n ASSUMPTIONS \\n =====================\")\n",
    "    print (\"1: Any Multi-mapped read has the Original Phyla in ATLEAST 1 alignment\")\n",
    "    print (\"2: Don't know what to do with Rose Sequence Ignoring for now\")\n",
    "    print (\"3: No reference of Eukaryotes added\")\n",
    "    print (\"OVERALL: Atmost 10% Reads could have been mapped more.\")\n",
    "    print (\"Eukaryotes Counts: {0}({1:.2f}%)\".format(euCount, euCount/totReads))\n",
    "    print (\"Rose Counts: {0}({1:.2f}%)\".format(roseCount, roseCount/totReads))\n",
    "    print (\"====================================================================================\")\n",
    "    print (\"\\n ===================== \\n ACCURACY METRIC \\n =====================\")\n",
    "    print (\"Number of True positives(TP) reads: {0} ({1:.2f}M, {2:.2f}%)\".format(TP, TP/mil, TP*hund/totReads))\n",
    "    print (\"Number of False Negatives(FN) reads: {0} ({1:.2f}M, {2:.2f}%)\".format(FN, FN/mil, FN*hund/totReads))\n",
    "    print (\"Number of False positives(FP) reads: {0} ({1:.2f}M, {2:.2f}%)\".format(FP, FP/mil, FP*hund/totReads))\n",
    "    print (\"Number of True Negatives(TN) reads: {0} ({1:.2f}M, {2:.2f}%)\".format(TN, TN/mil, TN*hund/totReads))\n",
    "    print (\"====================================================================================\")\n",
    "            \n",
    "get_stats(\"../bam/A1.sam\", \"../reads/A1_1.fastq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
